pkg/ekcoops/poll.go

```diff
@@ -2,34 +2,140 @@ package ekcoops
 
 import (
 	"context"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/json"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
 	"time"
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
 // Poll will run the reconcile function on an interval
 func (o *Operator) Poll(ctx context.Context, interval time.Duration) {
+	logger := o.controller.Log
 	ticker := time.NewTicker(interval)
 
+	dynamicClient, err := dynamic.NewForConfig(o.controller.Config.ClientConfig)
+	if err != nil {
+		logger.Error("Failed creating Kubernetes client.", err)
+	}
+
+	installerList, err := dynamicClient.Resource(schema.GroupVersionResource{
+		Group:    "cluster.kurl.sh",
+		Version:  "v1beta1",
+		Resource: "installers",
+	}).
+		List(metav1.ListOptions{})
+	if err != nil {
+		logger.Error("Failed to fetch.", err)
+	}
+
+	var (
+		nodeWatcher        watch.Interface
+		prometheusClient   dynamic.ResourceInterface
+		alertManagerClient dynamic.ResourceInterface
+	)
+
+	if len(installerList.Items) != 0 {
+		var currentInstaller unstructured.Unstructured
+		for _, installer := range installerList.Items {
+			selectedInstallerTimestamp := currentInstaller.GetCreationTimestamp()
+			nextInstallerTimestamp := installer.GetCreationTimestamp()
+			if (&selectedInstallerTimestamp).Before(&nextInstallerTimestamp) {
+				currentInstaller = installer
+			}
+		}
+
+		prometheusClient = dynamicClient.Resource(prometheusGvr).Namespace("monitoring")
+		alertManagerClient = dynamicClient.Resource(alertManagerGvr).Namespace("monitoring")
+
+		prometheus, _ := prometheusClient.Get("k8s", metav1.GetOptions{})
+		alertManager, _ := alertManagerClient.Get("prometheus-alertmanager", metav1.GetOptions{})
+
+		if prometheus != nil && alertManager != nil {
+			nodeWatcher, err = o.controller.Config.Client.CoreV1().Nodes().Watch(metav1.ListOptions{})
+			if err != nil {
+				logger.Error("Failed to start watcher on nodes.", err)
+			}
+		}
+	}
+
 	i := 0
 	for {
 		select {
 		case <-ticker.C:
 			nodeList, err := o.client.CoreV1().Nodes().List(metav1.ListOptions{})
 			if err != nil {
-				o.log.Infof("Skipping reconcile: failed to list nodes: %v", err)
+				logger.Infof("Skipping reconcile: failed to list nodes: %v", err)
 				continue
 			}
 			doFullReconcile := i%60 == 0
 			err = o.Reconcile(nodeList.Items, doFullReconcile)
 			if err != nil {
-				o.log.Infof("Reconcile failed: %v", err)
+				logger.Infof("Reconcile failed: %v", err)
 				continue
 			}
+		case event := <-nodeWatcher.ResultChan():
```

Comment:
do we need to check when channel is closed? `event, ok := <-nodeWatcher.ResultChan()` and if `!ok` create new watcher? why not just use node list from ticker polling?

Reply:
+1 on listing the nodes from the ticket polling